man/dbdlbl.1
.TH DBDLBL 1 "12 September 1995"
.FB
.SH NAME
dbdlbl \- define a \fIFirstBase\fP label dictionary
.SH SYNOPSIS
.B dbdlbl
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbdlbl
allows definition and modification of a 
.I FirstBase 
label dictionary via a screen oriented editor.
The label dictionary created
is later used by 
\fIdblgen\fP(1)
to create a ready to spool print image file of the labels.
.PP
A label dictionary consists of valid database field_names
and other characters or words
plugged into slots of a 5 line label template. 
.PP
The user has free choice of design of the label, meaning any of the 
label's five lines can be used for any purpose.
Oftentimes, a certain line is chosen to be the
.I name_line
of the label and another for the 
.I address_line,
etc, all the way down to the zip code.
A
.I header_line
can also be specified to print at the top of each label.
.PP
.B Note:
it is not necessary to use 
\fIdbdlbl\fP(1) 
and
\fIdblgen\fP(1)
strictly for mailing labels. Another idea is to use these programs to
generate name tags or badges for conferences or meetings.
.PP
When creating a new label, \fIdbdlbl\fP(1) starts up in a forced automatic
add mode, with line one (1) being the starting point. The standard end key
signal can be used at this point to escape from this forced mode to
the 'any change' screen.
The header line, line zero (0), is not input
through this forced input mode.
.PP
All lines of the label consists of words.  More than one word may appear
on each line.
If a word is a valid 
.I field_name
from the current database dictionary, then it is accepted with 
no messages. 
.PP
If however, the word is not defined as a field_name in the dictionary, 
the user is notified that the line contains at least one undefined symbol
by a message displayed next to the label line.
The undefined symbol is also marked by a 
.I '^'
printed below the first character in the symbol.
.PP
During the course of label generation, the defined words from the label
template created using
\fIdbdlbl\fP(1)
are replaced with their database values.
Undefined symbols are output exactly as seen.
Note that long symbols are silently
truncated to fit into a database fieldname descriptor.
.B Note:
All white space is left intact, although entire lines that are
blank after any substitutions in the label generator phase are
collapsed out of the label.
.PP
Optionally, a file name to store the print image of the labels can be
input, as well as the number of labels across the top (1-4 up), the
size in characters per label, and whether or not to left justify the
header line (line 0).
.PP
A list of simple editor commands appears on the 24th line of the 
screen at all times, with the standard help signal (<CTL>-H) displaying a
small summary of the possible commands.
The following is a list of valid 
\fIdbdlbl\fP(1) commands:
.TP 7
.B <CTL>-H
displays a summary of these commands.
.TP 7
.B ?[dfhiN]
displays help on the database, formulas,
user defined help files, index, or field N.
.TP 7
.B #
allow changes to line #.
.TP 7
.B -
standard END key \(em exits program and keeps all changes.
.TP 7
.B <CTL-X> 
aborts the program. The effect is to exit without
overwriting the file, meaning all changes are lost.
.PP
Upon successful completion of a label dictionary, the user is given a
chance to run the label generator. All arguments given
to \fIdbdlbl\fP(1) are passed on to \fIdblgen\fP(1).
The processes are overlaid,
meaning no new shells are generated in using this option.
.SH FILES
.PD 0
.TP 10
cdblbl
default name for \fIFirstBase\fP labels.
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.idictl
dictionary for \fIFirstBase\fP labels.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.TP 10
*.lbl
\fIFirstBase\fP label print image file.
.PD
.SH SEE ALSO
dblgen(1), input(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbdlbl\fP(1)
will allow modification to any label dictionary file at any time. These
label dictionary files should thus be write protected if they are not to be
changed.
.PP
The defaults for the various optional fields reflect 4 up labels printed
using 12 cpi (pitch) on 54 character wide labels.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dblgen/dbdlbl.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbdlbl.c,v 9.1 2001/02/16 19:43:22 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbdlbl.c	8.2 11/26/93 FB";
#endif

/* 
 *  dbdlbl.c - screen editor for label dict file : idictl 
 */

#include <fb.h>
#include <fb_vars.h>
#include <dbd.h>

#define HLP_DBDLBL 	"dbdlbl.hlp"

#define COLUMN 15			/* column position of fb_input */
#define FORMAT 4			/* default format "up" */
#define ASIZE 54			/* standard label size at 12cps */
#define SLPRT "cdblbl"			/* standard base output file name */

extern short int cdb_locklevel;
extern short int cdb_askgen;
extern char *cdb_LGEN;

static char filen[FB_MAXNAME] = {""};	/* name of label dictionary */
static char rcfile[FB_MAXNAME] = {""};	/* name of output label file */
static char justify = 'n';		/* left justify header? flag */

static int fmt = 0;			/* format 1-4 up */
static int asize = 0;			/* label size in characters (cpi) */

static int anychange();

struct lab {				/* for a each lab line */
   char aid[FB_SCREENFIELD+1];
   };

static fb_database *hp;

static ed_labrc();
static fput();
static void check();
static iskey();
static fget();
static output();
static headers();

/* 
 * dbdlbl - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
	 cdb_locklevel = -1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_initdbd(hp);
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);
	 fb_scrhdr(hp, "Define Labels");
         if (ed_labrc(argv) != FB_END)
	    fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 if (cdb_askgen)
	    fb_chainto("Generate Labels", cdb_LGEN, argv);
	 fb_ender();
      }
      
/* 
 *  ed_labrc - edit a label dictionary file 
 */
 
   static ed_labrc(argv)
      char *argv[];

      {
	 fb_field *keyprint[FB_MAXKEEP];
	 struct lab labels[5];
	 char abuf[FB_MAXLINE];
	 int i, p, st;

         initlabel(cdb_keymap, keyprint, hp,
                        filen, argv, hp->idict, &fmt, &asize);
	 headers();
	 sprintf(rcfile, "%sl", hp->idict);
	 if (fmt == 0){				/* no dictl file */
	    strcpy(filen, SLPRT);
	    fmt = FORMAT;
	    asize = ASIZE;
	    for (i = 0; i <= 4; i++)
	       labels[i].aid[0] = NULL;
	    for (i = 0; i <= 8; i++)
	       fput(labels, i);
	    for (i = 1; i <= 4; i++)
	       if (fget(labels, i) == FB_END)
		  break;
	    }
	 else{
	    if (fmt < 0){
	       fmt = -fmt;
	       justify = 'n';
	       }
	    else
	       justify = 'y';
	    for (p = 0, i = 0; ; p++){
	       abuf[0] = NULL;
	       while (keyprint[i] != NULL)
		  strcat(abuf, keyprint[i++]->id);
	       if (p >= 5)
		  fb_xerror(FB_MESSAGE, "too many label lines", NIL);
	       strcpy(labels[p].aid, abuf);
	       ++i;
	       if (keyprint[i] != NULL && keyprint[i]->size == 0)
		  break;
	       }
	    for (i = 0; i <= 8; fput(labels, i++))  /* 1-4=lines;5+=other */
	       ;
	    }
	 st = anychange(labels);
	 if (st == FB_END)
	    output(labels);
	 return(st);
      }

/* 
 *  fput - fb_put an entire label line to a constant screen location 
 */
 
   static fput(k, i)
      struct lab k[];
      int i;
      
      {
         int row;
         
	 check(k, i);
         row = (i + 1) * 2 + 3;
	 if (i >= 5)
	    row++;
         fb_move(row, COLUMN);
	 if (i >= 0 && i <= 5)
	    fb_printw("%2d> ", i);
	 else
	    fb_printw("    ");
         switch (i){
            case 0:
	       if (justify == 'n')
	          fb_printw("%50s", k[i].aid);
	       else 
	          fb_printw("%-50s", k[i].aid);
	       break;
            case 1: case 2: case 3: case 4: 
	       fb_printw("%-50s", k[i].aid); break;
            case 5: 
	       fb_printw("Output to: ---> %s", filen); break;
            case 6: 
	       fb_printw("Label Format:   %d up", fmt); break;
            case 7: 
	       fb_printw("Label Size:     %d char/label", asize); break;
            case 8: 
	       fb_printw("Left Justify 0? %c ", justify); break;
            }
      }

/* 
 *  check - check lines of label for constants -- mark them with '^'
 */
 
   static void check(k, i)
      struct lab k[];
      int i;

      {
	 char line[FB_MAXLINE], word[FB_MAXLINE], out[FB_MAXLINE], *fb_trim();
	 int j, p, undef, row;

	 if (i < 0 || i > 4)
	    return;
         row = (i + 1) * 2 + 4;
	 undef = 0;
	 out[0] = NULL;
	 line[0] = NULL;
	 sprintf(line, "%50s", "");
	 for (p = 1, j = 1; (j = fb_gettoken(k[i].aid, j, word, '_')) != 0; ){
	    if (iskey(word) != 0){
	       line[p-1] = '^';
	       undef = 1;
	       }
	    if (strlen(word) > FB_TITLESIZE)	/* in lgen, words are */
	       word[FB_TITLESIZE] = NULL;		/* stored in fields */
	    strcat(out, word);
	    p = strlen(out) + 1;
	    }
	 strcpy(k[i].aid, out);
	 fb_move(row, 1);
	 if (undef == 0)
	    fb_printw("%13s", " ");
	 else
	    fb_printw("undefined -->");	/* len = 13 */
	 if (i == 0 && (justify == 'n' || justify == 'N')){
	    p = strlen(out);
	    line[p] = NULL;
	    sprintf(out, "%50s", line);
	    }
	 else
	    strcpy(out, line);
	 fb_move(row, COLUMN + 4);
	 fb_printw(FB_FSTRING, out);
      }

/* 
 *  iskey - determine if word is a fb_field key name or ignorrable (ie white) 
 */
 
   static iskey(word)
      char *word;

      {
	 int i;

         if ((word[0] >= 'A' && word[0] <= 'Z') ||
	     (word[0] >= 'a' && word[0] <= 'z') ||
	     (word[0] >= '0' && word[0] <= '9')){
	    for (i = 0; i < hp->nfields; i++)
	       if (strcmp(word, cdb_kp[i]->id) == 0)
		  return(0);
	    return(1);
	    }
	 else
	    return(0);		/* not worth noting: ignore NON-FB_ALPHA */
      }

/* 
 * fget - get an entire label line from a constant screen location 
 */
 
   static fget(k, i)
      struct lab k[];
      int i;
      
      {
         int row, st, num, xend;
	 char abuf[FB_MAXLINE], *fb_trim();
         
         row = (i + 1) * 2 + 3;
	 if (i >= 5)
	    row++;
	 fb_fmessage("Enter Label Line, <CTL>-X = abort");
         switch (i){
            case 0: case 1: case 2: case 3: case 4: 
	       if (i != 1)
		  xend = FB_NOEND;
	       else
		  xend = FB_OKEND;
	       st = fb_input(row, COLUMN + 4, FB_SCREENFIELD, 0,
	          FB_ALPHA, abuf, FB_ECHO, xend, FB_CONFIRM);
	       if (st == FB_END){
		  fput(k, 1);
		  return(FB_END);
		  }
	       if (st == FB_DEFAULT)
		  k[i].aid[0] = NULL;
	       else if (st == FB_AOK)
		  strcpy(k[i].aid, fb_trim(abuf));
	       break;
            case 5:
	       /* 
	        *  get file name 
		*/
	       st = fb_input(row, COLUMN+20 , FB_TITLESIZE, 0, FB_ALPHA,
	          abuf, FB_ECHO, FB_NOEND, FB_CONFIRM);
	       if (st == FB_DEFAULT)
		  strcpy(filen, SLPRT);
	       else if (st == FB_AOK){
		  strcpy(filen, fb_trim(abuf));
		  }
	       fput(k, i++);
	       
	       /* 
	        *  get label format 
		*/
		
	       row += 2;
	       st = fb_input(row, COLUMN+20 , 1, 0, FB_INTEGER,
	          (char *) &num, FB_ECHO, FB_NOEND, FB_CONFIRM);
	       if (st == FB_DEFAULT || (st == FB_AOK && (num < 1 || num > 4)))
		  fmt = 4;
	       else if (st == FB_AOK)
		  fmt = num;
	       fput(k, i++);
	       
	       /* 
	        *  get label size  -- char/label 
		*/
		
	       row += 2;
	       st = fb_input(row, COLUMN+20 , 2, 0, FB_INTEGER,
	          (char *) &num, FB_ECHO, FB_NOEND, FB_CONFIRM);
	       if (st == FB_DEFAULT || (st == FB_AOK && (num < 1 || num > 220)))
		  asize = ASIZE;
	       else if (st == FB_AOK)
		  asize = num;
	       fput(k, i++);

	       /* 
	        *  justify header flag 
		*/
		
	       row += 2;
	       st = fb_input(row, COLUMN+20 , 1, 0, FB_ALPHA,
	          abuf, FB_ECHO, FB_NOEND, FB_CONFIRM);
	       if (st == FB_DEFAULT)
		  justify = 'n';
	       else if (st == FB_AOK){
	          if (justify != 'y' && justify != 'n' &&
	              justify != 'Y' && justify != 'N')
		     abuf[0] = 'n';
		  justify = abuf[0];
		  }
	       if (st != FB_ABORT)
	          fput(k, 0);
	       break;
            }
	 fput(k, i);
	 
	 /* 
	  *  finally return
	  */
	  
	 return(FB_AOK);
      }
            
/* 
 *  anychange - set up anychange screen 
 */
 
   static anychange(labels)
      struct lab labels[];
      
      {
         char inp[6];
         int st, num, i;
         
	 for(;;){
	    fb_fmessage("Any Change (#, <CTL>-H=Help, -=End) ? ");
	    st = fb_input(-cdb_t_lines, 40, 4, 0, FB_ALPHA, inp, FB_ECHO,
               FB_OKEND, FB_CONFIRM);
	    				/* allow FB_QHELP signal */
            if (st == FB_END)
	       break;
	    else if (st == FB_DEFAULT)
	       continue;
            else if (st == FB_ABORT){
	       if (fb_mustbe('y', "Really Quit? (y/n) ",
                  cdb_t_lines, 1) == FB_AOK)
		     break;
		  }
	    else if (st == FB_AOK){
	       if (inp[0] == FB_HELP)
		  fb_help(inp, hp);
	       else {
	          if (isdigit(inp[0])){
		     num = atoi(inp);
		     if (num < 0 || num > 5)
			fb_screrr("Invalid Selection");
		     else
			fget(labels, num);
		     }
		  else 
		     fb_screrr("Invalid Selection");
		  continue;	/* break out so as not to redraw screen */
		  }
	       }
	    else if (st == FB_QHELP)
	       fb_fhelp(HLP_DBDLBL);
	    fb_move(2, 1);
	    fb_clrtobot();
	    headers();
	    for(i = 0; i <= 8; i++)
	       fput(labels, i);
	    }
	 return(st);
      }
      
/* 
 * output - output the label dictionary 
 */
 
   static output(k)
      struct lab k[];

      {
	 int i;
	 FILE *fs, *fb_mustfopen();

	 fs = fb_mustfopen(rcfile, "w");
	 for (i = 0; i < 5; i++)
	    fprintf(fs, "%s\n", k[i].aid);
	 if (justify == 'n' || justify == 'N')
	    fmt = -fmt;
	 fprintf(fs, "%% %d %d\n", fmt, asize);
	 fprintf(fs, "%s\n", filen);
	 fclose(fs);
      }

/* 
 * headers - draw headliner box 
 */
 
   static headers()
      {
	 int i;

	 fb_move(4, COLUMN); 
	 fb_printw("-------------------------------------------------------");
	 fb_move(15, COLUMN); 
	 fb_printw("-------------------------------------------------------");
	 for (i = 4; i <= 15; i++){
	    fb_move(i, COLUMN-1);
	    if (i == 4)
	       fb_printw("/");
	    else if (i == 15)
	       fb_printw("\\");
	    else
	       fb_printw("|");
	    }
	 for (i = 4; i <= 15; i++){
	    fb_move(i, COLUMN+55);
	    if (i == 4)
	       fb_printw("\\");
	    else if (i == 15)
	       fb_printw("/");
	    else
	       fb_printw("|");
	    }
	 fb_infoline();
      }

/* 
 *  usage fb_fmessage 
 */
 
   usage()
      
      {
         fb_xerror(FB_MESSAGE, "usage: dbdlbl [-d dbase] [-i index]", NIL);
      }
